Convert Bodhi to publish messages via AMQP

jeremycline opened the pull request:
This converts Bodhi to use fedora-messaging to publish to an AMQP message broker. It also sets up a RabbitMQ message broker in the Vagrant environment.

This pull request does the minimal amount necessary to publish via AMQP. It does not define message schema, nor does it attempt to convert the existing fedmsg consumers to the fedora-messaging consumer API. These will be done in follow-up pull requests.

jeremycline committed:
Add RabbitMQ to the development environment

In preparation for the transition from fedmsg to fedora-messaging, set
up an AMQP broker in the development environment and print all messages
sent to it to the journal like fedmsg-tail.

Signed-off-by: Jeremy Cline <jcline@redhat.com>

jeremycline committed:
Add message publishing via fedora_messaging

This does the bare minimum to move Bodhi from fedmsg to
fedora_messaging. It replaces the call to ``fedmsg.publish`` with
``fedora_messaging.api.publish`` and sends all messages as plain
``fedora_messaging.api.Message`` objects rather than defining schema.

It leaves the fedmsg publishing in place, controlled by the
"fedmsg_enabled" configuration, so it is possible to switch back to
fedmsg if necessary.

All the unit tests have been adjusted so it mocks out the publishing
call and asserts messages are sent. It does not, however, make strong
assertions about the type or content of the message since they are all
sent as plain ``fedora_messaging.api.Message`` objects. I plan to send a
follow-up PR with the start of the schema definitions and as schema are
defined and used, the tests can be updated to make stronger assertions.

Signed-off-by: Jeremy Cline <jcline@redhat.com>

bowlofeggs commented:
I'm going to rebase this bad boy because the f27 tests are no longer run and don't work.

bowlofeggs commented on devel/ansible/roles/bodhi/tasks/main.yml:
```diff
@@ -90,6 +91,7 @@
       - python3-pylibravatar
       - python3-pyramid
       - python3-pyramid-fas-openid
+      - python3-pyramid-tm
```
I am able to run the tests without pyramid-tm, but I did have to remove it from my ```development.ini``` file. Is it possible you need to refresh your ```development.ini``` file from ```devel/development.ini.example```? You can just delete it and re-vagrant-up it.

Related: https://github.com/fedora-infra/bodhi/pull/2755

bowlofeggs commented on Vagrantfile:
```diff
@@ -20,6 +20,10 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
 
+ # Forward traffic on the host to the RabbitMQ management UI on the guest.
+ # This allows developers to view message queues at http://localhost:15672/
+ config.vm.network "forwarded_port", guest: 15672, host: 15672
```
We should document the default username/password in ```docs/developer/vagrant.rst```.

bowlofeggs commented on bodhi/server/notifications.py:
```diff
@@ -97,7 +98,68 @@ def send_fedmsgs_after_commit(session):
+    """
+    Send a message via AMQP or fedmsg.
+
+    This use used to send a message to fedmsg if "fedmsg_enabled" is true in the
```
```suggestion
    This is used to send a message to fedmsg if "fedmsg_enabled" is true in the
```

bowlofeggs commented on bodhi/server/notifications.py:
```diff
@@ -97,7 +98,68 @@ def send_fedmsgs_after_commit(session):
+    # Dirty, nasty hack that I feel shame for: fedmsg modifies messages quietly
+    # if they have objects with __json__ methods on them. For now, copy that
+    # behavior. In the future, callers should pass fedora_messaging.api.Message
+    # sub-classes or this whole API should go away.
```
haha, I guess I should file a ticket to switch to this format too.

jeremycline replied:
Yeah, I figured doing it in bits and pieces would be easiest. I don't mind filing tickets for the remaining work items (defining schema, using them, moving consumers over to ``fedora-messaging consume`` from fedmsg-hub)

bowlofeggs commented on bodhi/server/notifications.py:
```diff
@@ -97,7 +98,68 @@ def send_fedmsgs_after_commit(session):
+    # that behavior here for now. In the future callers should pass proper topics.
+    fedmsg_conf = fedmsg.config.load_config()
+    topic = "{prefix}.{env}.bodhi.{t}".format(
+        prefix=fedmsg_conf["topic_prefix"], env=fedmsg_conf["environment"], t=topic)
```
I wonder if I should just fix this with Bodhi 4 by declaring that Bodhi's message topic prefixes are changing? Would you advise that I just remove the prefix and env and just start the topic with "bodhi"? Or maybe we could make the topic prefix be a Bodhi setting in Bodhi 4 instead of loading the fedmsg config. It'd be nice to fully drop fedmsg in Bodhi 4.

jeremycline replied:
It would be very nice to fix the topics, but I think maybe the best migration path is something like:

1. Add new message types with reasonable topics while still publishing old messages with old topics.
2. Publish both old and new messages.
3. Make an announcement about the old messages going away, convert all known consumers of those messages to the new message
4. Drop the old messages.

There's not really anything in fedora-messaging to help you in this, although if we find a common pattern that most apps would use we should definitely add that.

As for what the topic should be, I'd recommend starting with "bodhi" and getting from more general to less general. Something like "bodhi.update.created.<dist>.<pkgname>" could work. Topics can (and in many cases should) be dynamic. There a few restrictions though - they need to be 255 characters (or octets, possibly) or less.

edit: and yeah, I thought about making it be a bodhi setting, if you'd prefer that I can make that happen in this PR. That way dropping fedmsg is simpler for Bodhi 4

bowlofeggs commented on bodhi/tests/server/consumers/test_masher.py:
```diff
@@ -2666,7 +2668,8 @@ def test_expires_compose_updates(self):
         t.id = getattr(self.db.query(Release).one(), '{}_tag'.format('stable'))
 
-        t.perform_gating()
+        with mock_sends(api.Message):
```
Ooooh that seems nice! You guys thought of everything.

bowlofeggs commented on bodhi/tests/server/test_notifications.py:
```diff
@@ -127,13 +128,28 @@ class TestPublish(base.BaseTestCase):
 
+    @mock.patch.dict('bodhi.server.config.config', {'fedmsg_enabled': False})
+    def test_fedmsg_publish_off_force(self, mock_init):
+        """Assert when fedmsg publish is off, fedora-messaging messages are queued."""
```
```suggestion
        """Assert that fedora-messaging messages respect the force flag."""
```

bowlofeggs commented on bodhi/tests/server/test_notifications.py:
```diff
@@ -127,13 +128,28 @@ class TestPublish(base.BaseTestCase):
+        expected = api.Message(topic='org.fedoraproject.dev.bodhi.demo.topic',
+                               body={u'such': 'important'})
+
+        with fml_testing.mock_sends(expected):
```
Man that's a nice API.

bowlofeggs commented on devel/ansible/roles/bodhi/files/print-messages.service:
```diff
@@ -0,0 +1,11 @@
+
+[Service]
+User=vagrant
+ExecStart=/usr/bin/fedora-messaging consume --callback="fedora_messaging.example:printer"
```
Might be cool to add a .bashrc alias like ```bmessaging``` or something that tails these logs. If you opt for that, don't forget to add it to the motd too ☺

bowlofeggs reviewed (changes requested):
This looks great!

Does fedora messaging have something like fedmsg-hub, or will we just convert fedmsg-hub to use fedora-messaging instead at some point in the future? I'm curious if we should do something to make the composer and updates handlers work with it.

bowlofeggs commented:
To get the tests to pass, you can add ```fedora-messaging``` to the Dockerfiles in ```devel/ci/```. For releases that don't have ```fedora-messaging``` (IIRC, it's only in F29+?), feel free to just pip install it.

bowlofeggs commented:
Oh, and if you want to run the CI tests locally to more quickly test your ```Dockerfile``` changes, check out the Vagrant alias ```bci --help```.

jeremycline committed:
Update the CI environment

Add fedora-messaging to the Dockerfiles, using pip when the package
isn't available.

Signed-off-by: Jeremy Cline <jcline@redhat.com>

bowlofeggs reviewed (approved):
Thanks!

bowlofeggs commented:
Hmm, mergify seems stuck. I'll rebase.

bowlofeggs commented:
@abompard It seems that the integration tests are failing here, but I am unsure why:

https://ci.centos.org/job/bodhi-pipeline/job/PR-2754/17/execution/node/99/log/

It does appear to be reproducible locally.

bowlofeggs commented:
Aha! It's buried in the middle of LOTS of output - looks like we just need to add fedora_messaging to the integration test container too. @jeremycline I'll take care of that for you.

bowlofeggs commented:
Hmm, now there seems to be failing unit tests. This may be due to rebasing on other patches, not sure yet. I'm investigating.

bowlofeggs commented:
I am applying this patch to fix the unit tests:

```
diff --git a/bodhi/tests/server/services/test_updates.py b/bodhi/tests/server/services/test_updates.py
index 79932ef0..0f20a6a5 100644
--- a/bodhi/tests/server/services/test_updates.py
+++ b/bodhi/tests/server/services/test_updates.py
@@ -3086,8 +3086,7 @@ class TestUpdatesService(BaseTestCase):
         publish.assert_called_with(topic='update.edit', msg=ANY)

     @mock.patch(**mock_valid_requirements)
-    @mock.patch('bodhi.server.notifications.publish')
-    def test_pending_update_on_stable_karma_reached_autopush_enabled(self, publish, *args):
+    def test_pending_update_on_stable_karma_reached_autopush_enabled(self, *args):
         """Ensure that a pending update stays in testing if it hits stable karma while pending."""
         nvr = u'bodhi-2.0.0-2.fc17'
         args = self.get_update(nvr)
```

bowlofeggs commented:
There is an integration test failure in CentOS CI, but I cannot reproduce when I run it locally. I'm going to try to bump your branch to get CI to run again and see if its repeatable.

bowlofeggs commented:
I have filed a bug about the integration test failure: https://github.com/fedora-infra/bodhi/issues/2835

bowlofeggs commented:
Ha, and now a conflict. I'll rebase again to resolve that. Who knows, maybe that'll get the tests to pass!

Commit:
Merge branch 'develop' into 'fedora-messaging'

bowlofeggs commented:
This feature is planned for inclusion in Bodhi 3.13: #2917.

bowlofeggs commented:
This feature is now deployed to staging: https://bodhi.stg.fedoraproject.org/docs/user/release_notes.html

bowlofeggs commented:
Bodhi 3.13.0 has been released:

https://github.com/fedora-infra/bodhi/releases/tag/3.13.0

Fedora Infrastructure plans to deploy it to production on Monday.
